### PR TITLE
gitlab: allow usage of advanced search feature

### PR DESCRIPTION
**Pull request by xanderio**
###### Description of changes

In order to enable and use the advanced search feature in GitLab EE, gitlab requires this program to be compiled and it's location specified in the gitlab.yml file.

We've successfully deployed this to our production GitLab installation and so I can confirm the produced binary works.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/gitlab.nix [hunk 1]
```diff
@@ -159,6 +159,7 @@ let
         api_url = "http://${config.services.dockerRegistry.listenAddress}:${toString config.services.dockerRegistry.port}/";
         issuer = cfg.registry.issuer;
       };
+      elasticsearch.indexer_path = "${pkgs.gitlab-elasticsearch-indexer}/bin/gitlab-elasticsearch-indexer";
```
Comment (teto) [inline nixos/modules/services/misc/gitlab.nix:162]: this unconditionnaly adds it to closure, cant it be passed via extraConfig instead ? or maybe it warrants a dedicated option

**Comment by yu-re-ka** [comment]
Is this component bound to a specific gitlab version, can we integrate it into the update script?

**Comment by xanderio** [comment]
We can pull the version from the `GITLAB_ELASTICSEARCH_INDEXER` file in the main gitlab repository. I'll add this to the update script.

File: pkgs/applications/version-management/gitlab/update.py [hunk 2]
```diff
@@ -201,6 +201,16 @@ def update_gitlab_workhorse():
...
 
 
+@cli.command('update-gitlab-elasticsearch-worker')
+def update_gitlab_elasticsearch_worker():
+    """Update gitlab-elasticsearch-worker"""
+    data = _get_data_json()
+    gitlab_elasticsearch_worker_version = data['passthru']['GITLAB_ELASTICSEARCH_INDEXER_VERSION']
+    _call_nix_update('gitlab-elasticsearch-worker', gitlab_elasticsearch_worker_version)
+
+    subprocess.check_output(['./update-vendorHash.sh'], cwd='gitlab-elasticsearch-worker')
```
Comment (yayayayaka) [inline pkgs/applications/version-management/gitlab/update.py:211]: [nix-update](https://github.com/Mic92/nix-update) already takes care of updating the vendor hash according to their [README](https://github.com/Mic92/nix-update/blob/master/README.md#features).

File: pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  vendorHash = "sha256-MyIK8qaSjkNIk1at3iRckiaOZ6+0M7WNE/nYTUA9gg8=";
+
+  buildInputs = [ icu ];
+  nativeBuildInputs = [ pkg-config ];
+
+  meta = with lib; {
+    description = "Indexes Git repositories into Elasticsearch for GitLab.";
+    license = licenses.mit;
+    maintainers = [ maintainers.xanderio ];
```
Comment (yayayayaka) [inline pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix:23]: Feel free to also add myself as maintainer :heart:

File: pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+buildGoModule rec {
+  pname = "gitlab-elasticsearch-indexer";
+  version = "4.3.0";
+  rev = "v${version}";
+
+  src = fetchFromGitLab {
+    owner = "gitlab-org";
+    repo = "gitlab-elasticsearch-indexer";
+    inherit rev;
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix:11]:
```suggestion

  src = fetchFromGitLab {
    owner = "gitlab-org";
    repo = "gitlab-elasticsearch-indexer";
    rev = "v${version}";
```

File: pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  vendorHash = "sha256-MyIK8qaSjkNIk1at3iRckiaOZ6+0M7WNE/nYTUA9gg8=";
+
+  buildInputs = [ icu ];
+  nativeBuildInputs = [ pkg-config ];
+
+  meta = with lib; {
+    description = "Indexes Git repositories into Elasticsearch for GitLab.";
+    license = licenses.mit;
+    maintainers = [ maintainers.xanderio ];
+    platforms = platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix:24]:
```suggestion
```
thats the default

File: pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/update-vendorHash.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+# Break the hash
+sed -i 's/vendorSha256.*$/vendorSha256 = "0000000000000000000000000000000000000000000000000000";/g' "${defaultnix}"
+
+# Try to build
+echo "Trying to build ${defaultnix}" >&2
+# Hash was wrong so get the proper one
+correctHash="$(grep got: <<<"$(nix-build ../../../../.. -A ${attr} 2>&1 || :)" | awk '{print $2}' | xargs nix hash to-base32)"
+# Add the hash into the file
+sed -i "s/vendorSha256.*$/vendorSha256 = \"${correctHash}\";/g" "${defaultnix}"
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/update-vendorHash.sh:17]:
```suggestion
```

File: pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/update-vendorHash.sh [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env nix-shell
+#! nix-shell -I nixpkgs=../../../../.. -i bash -p bash
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/update-vendorHash.sh:2]:
```suggestion
#! nix-shell -I nixpkgs=../../../../.. -i bash
```

File: pkgs/applications/version-management/gitlab/update.py (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/applications/version-management/gitlab/update.py:211]: That is correct

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix
Comment (yayayayaka) [inline pkgs/applications/version-management/gitlab/gitlab-elasticsearch-worker/default.nix:1]: Rename this file to `gitlab-elasticsearch-indexer.nix`

File: pkgs/applications/version-management/gitlab/update.py [hunk 8]
```diff
@@ -342,6 +351,7 @@ def commit_gitlab(old_version: str, new_version: str, new_rev: str) -> None:
             "gitlab-pages",
             "gitlab-shell",
             "gitlab-workhorse",
+            "gitlab-elasticsearch-worker",
```
Comment (yayayayaka) [inline pkgs/applications/version-management/gitlab/update.py:354]:
```suggestion
            "gitlab-elasticsearch-indexer",
```

File: pkgs/applications/version-management/gitlab/update.py [hunk 9]
```diff
@@ -297,6 +297,14 @@ def update_gitlab_container_registry(rev: str, commit: bool):
         )
 
 
+@cli.command('update-gitlab-elasticsearch-worker')
+def update_gitlab_elasticsearch_worker():
+    """Update gitlab-elasticsearch-worker"""
+    data = _get_data_json()
+    gitlab_elasticsearch_worker_version = data['passthru']['GITLAB_ELASTICSEARCH_INDEXER_VERSION']
+    _call_nix_update('gitlab-elasticsearch-worker', gitlab_elasticsearch_worker_version)
```
Comment (yayayayaka) [inline pkgs/applications/version-management/gitlab/update.py:305]:
```suggestion
@cli.command('update-gitlab-elasticsearch-indexer')
def update_gitlab_elasticsearch_indexer():
    """Update gitlab-elasticsearch-indexer"""
    data = _get_data_json()
    gitlab_elasticsearch_indexer_version = data['passthru']['GITLAB_ELASTICSEARCH_INDEXER_VERSION']
    _call_nix_update('gitlab-elasticsearch-indexer', gitlab_elasticsearch_indexer_version)
```

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -8130,6 +8130,8 @@ with pkgs;
 
   gitlab-workhorse = callPackage ../applications/version-management/gitlab/gitlab-workhorse { };
 
+  gitlab-elasticsearch-worker = callPackage ../applications/version-management/gitlab/gitlab-elasticsearch-worker { };
```
Comment (yayayayaka) [inline pkgs/top-level/all-packages.nix:8133]: The upstream repository is called `gitlab-elasticsearch-indexer`

```suggestion
  gitlab-elasticsearch-indexer = callPackage ../applications/version-management/gitlab/gitlab-elasticsearch-indexer { };
```

**Review by yayayayaka — CHANGES_REQUESTED** [review]
Looks like you have been mixing up the package name. In the module, you are referencing the package as `pkgs.gitlab-elasticsearch-indexer` whereas in `pkgs/` you define the package as `gitlab-elasticsearch-worker`.

I suggest renaming the package to `gitlab-elasticsearch-indexer`.

**Comment by xanderio** [comment]
@yayayayaka I've renamed the package as suggested.

**Comment by yu-re-ka** [comment]
There are still some mentions of gitlab_elasticsearch_worker in the update script

**Comment by xanderio** [comment]
Should be all fixed now.

**Pull request merged**